### PR TITLE
Query caching

### LINKED ISSUE #86: Query and model caching
Request mechanism to accelerate inference by caching and reusing models and previously answered queries.

Cache should be persistent through restarts of the service

### PR DESCRIPTION
**Pull request by chrlen**
Resolves #86 

Add caching functionality by contributing multiple cache-classes and benchmarks to evaluate impact on performance.

### CODE REVIEW

File: mb_modelbase/cache/BaseCache.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+class BaseCache(abc.ABC):
+    """
+    Abstract baseclass for modelbase caches
+
+    Override function _get and _set to use various storage backends
+
+    Attributes:
+        _serialize (bool): sets if payload gets serialized before storage
+
+    """
```
Comment (nandaloo) [inline mb_modelbase/cache/BaseCache.py]: Adhere to PEP 257 (https://www.python.org/dev/peps/pep-0257/).
In particular but not only: (1) leading """ is not on a separate line. (2) Trailing """ is not not preceeded by an entirely empty line, and (3) docstrings should be complete sentences.

File: mb_modelbase/cache/DictCache.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import time
+import pickle
+import threading
+
+
+class DictCache(BaseCache):
+    """
+    Cache with a regular python dictionary as a backend.
+    Stores the content after every save_interval to a file on disk
+    Ensures mutual exclusion with a lock
```
Comment (nandaloo) [inline mb_modelbase/cache/DictCache.py]: Missing trailing "."

File: mb_modelbase/cache/DictCache.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    Attributes:
+        save_path (str): path to where the cache should save the dictionary
+        save_interval (float): interval in seconds on which to save the content of the dictionary
+
+    Todo:
+        Implement cache as a monitor like in the readers writers problem
+        Maybe lift mutual exclusion to superclass?
+    """
+
+    def __init__(self, save_path='/tmp/modelbaseCache', save_interval=30):
```
Comment (nandaloo) [inline mb_modelbase/cache/DictCache.py]: What if folder /tmp does not exist?

Comment (nandaloo) [reply mb_modelbase/cache/DictCache.py]: Alos, add missing documentation for arguments.

File: mb_modelbase/cache/MemcachedCache.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+from mb_modelbase.cache import BaseCache
+from pymemcache.client import base
+
+class MemcachedCache(BaseCache):
+    """
+    A cache with memcached as storage backend
+
+    Attributes:
+        hostname (str): hostname of memcached server
+        port (int): port of memcached server
```
Comment (nandaloo) [inline mb_modelbase/cache/MemcachedCache.py]: Why is hostname and port listed as attributes of the class if it really is paramters to the constructor?

File: mb_modelbase/server/modelbase.py [hunk 5]
```diff
@@ -147,7 +151,7 @@ class ModelBase:
             .float_format : The float format used to encode floats in a result. Defaults to '%.5f'
     """
 
-    def __init__(self, name, model_dir='data_models', load_all=True, watchdog=True):
+    def __init__(self, name, model_dir='data_models', load_all=True, cache=None, watchdog=True):
         """ Creates a new instance and loads models from some directory. """
```
Comment (nandaloo) [inline mb_modelbase/server/modelbase.py]: Please add documentation. I know it's not all your code, so thanks :)

File: mb_modelbase/server/modelbase.py [hunk 6]
```diff
@@ -267,6 +273,15 @@ def list_models(self):
...
 
     def execute(self, query):
+        path = ""
+        queryLogName = "interaction" + \
+                       time.strftime("%b:%d:%Y_%H", time.gmtime(time.time())) + ".log"
+
+        if not "SHOW" in query.keys():
+            with open(queryLogName, "a") as f:
+                f.write(json.dumps(query) + '\n')
+                logger.info(json.dumps(query))
```
Comment (nandaloo) [inline mb_modelbase/server/modelbase.py]: Is there a way to disable logging? it is not always necessary, as it really is just logging, correct? 
Please add a way to disable it, and make it the default.

File: mb_modelbase/server/modelbase.py [hunk 7]
```diff
@@ -290,15 +305,35 @@ def execute(self, query):
...
+
+            key = model_key(
+                name=query["AS"],
                 model=self._extractModel(query),
-                where=self._extractWhere(query),
-                default_values=self._extractDefaultValue(query),
-                default_subsets=self._extractDefaultSubset(query),
-                hide=self._extractHide(query)),
+                where=self._extractWhere(query)
+            )
```
Comment (nandaloo) [inline mb_modelbase/server/modelbase.py]: Reading this makes we wonder if the computation of the key should be part of the cache submodule?

File: mb_modelbase/server/modelbase.py [hunk 8]
```diff
@@ -290,15 +305,35 @@ def execute(self, query):
...
+
+            if self.cache is not None:
+                derived_model = self.cache.get(key)
+            else:
+                derived_model = None
+
+            if derived_model == None:
```
Comment (nandaloo) [inline mb_modelbase/server/modelbase.py]: test it with "is None"

File: mb_modelbase/server/modelbase.py [hunk 9]
```diff
@@ -290,15 +305,35 @@ def execute(self, query):
...
+                # maybe copy
+                derived_model = base if base.name == query["AS"] else base.copy(query["AS"])
+                # derive submodel
+                derived_model.model(
+                    model=self._extractModel(query),
+                    where=self._extractWhere(query),
+                    default_values=self._extractDefaultValue(query),
+                    default_subsets=self._extractDefaultSubset(query),
+                    hide=self._extractHide(query)),
+                if self.cache != None:
```
Comment (nandaloo) [inline mb_modelbase/server/modelbase.py]: ... and "is not None"

File: mb_modelbase/server/modelcache.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    res = [memCache.get(str(x)) for x in vals]
+
+    # docker run -p 6379:6379 --name lumen_redis_cache -d redis
+    redisCache = RedisCache({
+        'hostname': 'localhost',
+        'port': 6379
+    })
+
+    [redisCache.set( str(x), str(x)) for x in vals]
+    res = [redisCache.get(str(x)) for x in vals]
```
Comment (nandaloo) [inline mb_modelbase/server/modelcache.py]: What is this for? Should it be in a test case?

File: mb_modelbase/server/modelcache.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
+import abc
+from pymemcache.client import base as memcBase
+import redis
+
+class Cache():
+    def __init__(self, *args, **kwargs):
```
Comment (nandaloo) [inline mb_modelbase/server/modelcache.py]: Please document where appropriate, also the subclasses below.

File: mb_modelbase/cache/BaseCache.py (hunk 1 above)
Comment (chrlen) [reply mb_modelbase/cache/BaseCache.py]: Changing (1), even though leading """ on a seperate line is in the allowed by PEP 257.
Fixed (2) and (3)

File: mb_modelbase/cache/DictCache.py (hunk 2 above)
Comment (chrlen) [reply mb_modelbase/cache/DictCache.py:10]: Done!

File: mb_modelbase/cache/MemcachedCache.py (hunk 4 above)
Comment (chrlen) [reply mb_modelbase/cache/MemcachedCache.py]: Fixed!

File: mb_modelbase/cache/DictCache.py (hunk 3 above)
Comment (chrlen) [reply mb_modelbase/cache/DictCache.py]: Added documentation and a check if the directory exists.

File: mb_modelbase/server/modelbase.py (hunk 5 above)
Comment (chrlen) [reply mb_modelbase/server/modelbase.py:168]: Done! No problem :-)

File: mb_modelbase/server/modelbase.py (hunk 8 above)
Comment (chrlen) [reply mb_modelbase/server/modelbase.py]: Fixed!

File: mb_modelbase/server/modelcache.py (hunk 10 above)
Comment (chrlen) [reply mb_modelbase/server/modelcache.py]: The whole file is not needed anymore and has been removed.

File: mb_modelbase/server/modelcache.py (hunk 11 above)
Comment (chrlen) [reply mb_modelbase/server/modelcache.py]: See above.

File: mb_modelbase/server/modelbase.py (hunk 6 above)
Comment (chrlen) [reply mb_modelbase/server/modelbase.py]: Done!

File: mb_modelbase/server/modelbase.py (hunk 7 above)
Comment (chrlen) [reply mb_modelbase/server/modelbase.py]: Right, they shouldn't be part of the cache submodule. The key related functions have been moved to the ModelBase class, where all the _extract... functions are located.

**Pull request merged**